File: 백준/Gold/17299. 오등큰수/오등큰수.py
import sys
input = sys.stdin.readline

n = int(input())
num_list = list(map(int, input().split()))

cnt = {}
for num in num_list:
	if num not in cnt:
		cnt[num] = 1
	else:
		cnt[num] += 1

stack = []
result = [-1] * n

for i in range(n):
	while stack and cnt[num_list[stack[-1]]] < cnt[num_list[i]]:
		result[stack.pop()] = num_list[i]
	stack.append(i)

print(*result)

File: 백준/Gold/17299. 오등큰수/오등큰수.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        StringTokenizer st = new StringTokenizer(br.readLine());

        int[] nums = new int[n];
        for (int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
        }

        int[] result = new int[n];
        Arrays.fill(result, -1);

        int[] stack = new int[n];
        int top = -1;

        for (int i = 0; i < n; i++) {
            int num = nums[i];
            while (top >= 0 && map.get(num) > map.get(nums[stack[top]])) {
                result[stack[top--]] = num;
            }
            stack[++top] = i;
        }

        StringBuilder sb = new StringBuilder(n * 3);
        for (int i = 0; i < n; i++) {
            sb.append(result[i]);
            if (i + 1 < n) sb.append(' ');
        }
        System.out.print(sb);
    }
}